### PR TITLE
HPCC-18629 Python cache variables allow dev overwrite

### PR DESCRIPTION
**Pull request by Michael-Gardner**
Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@xwang2713 Please review.  This includes the system/platform.h changes that you made.

File: system/include/platform.h [hunk 1]
```diff
@@ -77,6 +77,14 @@
...
 #endif
 #include <endian.h>
+
+#ifdef _POSIX_C_SOURCE
+  #undef _POSIX_C_SOURCE
+#endif
+#ifdef _XOPEN_SOURCE
+  #undef _XOPEN_SOURCE
+#endif
+
```
Comment (mckellyln) [inline system/include/platform.h]: Why are these needed to be undef'ed ?

**Comment by Michael-Gardner** [comment]
@xwang2713 I honestly think we should just comment out those lines and add a note line that it needs to be uncommented if used in that special case.

**Comment by Michael-Gardner** [comment]
@mckellyln Is this fine for the platform.h file?  The commented out code in that file is for an edge case that we don't really want to document.

File: system/include/platform.h [hunk 2]
```diff
@@ -77,6 +77,15 @@
 #define _GNU_SOURCE
 #endif
 #include <endian.h>
+
+// Used for python 2.7 builds on cecntos 6
```
Comment (mckellyln) [inline system/include/platform.h]: centos spelling

File: system/include/platform.h [hunk 3]
```diff
@@ -77,6 +77,15 @@
...
+//#ifdef _POSIX_C_SOURCE
+//#undef _POSIX_C_SOURCE
+//#endif
+//#ifdef _XOPEN_SOURCE
+//#undef _XOPEN_SOURCE
+//#endif
+
```
Comment (mckellyln) [inline system/include/platform.h]: Should this really be for a certain version of Python and perhaps the compiler ?
(I understand that the OS often dictates the default version of Python and compiler but to be accurate should the comment really be for a specific version of Python ... ?
Could this logic/check/comment go into a pyembed.cpp file instead ?

Comment (Michael-Gardner) [reply system/include/platform.h]: I'll look into moving it into the pyembed code, that's a great idea.

**Comment by Michael-Gardner** [comment]
@mckellyln Can you please take a look at pyembed.cpp?  I went ahead and included a check on the python version, and then undefining the necessary values.

File: system/include/platform.h [hunk 4]
```diff
@@ -77,6 +77,7 @@
 #define _GNU_SOURCE
 #endif
 #include <endian.h>
+
```
Comment (richardkchapman) [inline system/include/platform.h]: Unnecessary change

File: plugins/pyembed/CMakeLists.txt [hunk 5]
```diff
@@ -37,13 +37,14 @@ if(PY2EMBED)
...
-    set(PYTHON2_DEBUG_LIBRARIES ${PYTHON_DEBUG_LIBRARIES})
-    set(PYTHON2LIBS_VERSION_STRING ${PYTHONLIBS_VERSION_STRING})
+
+    set(PYTHON2LIBS_FOUND ${PYTHONLIBS_FOUND} CACHE BOOL "Found Python2 Libraries")
+    set(PYTHON2_LIBRARY ${PYTHON_LIBRARY} CACHE PATH "Library")
+    set(PYTHON2_LIBRARIES ${PYTHON_LIBRARIES} CACHE PATH "Libraries")
+    set(PYTHON2_INCLUDE_PATH ${PYTHON_INCLUDE_PATH} CACHE PATH "include path")
+    set(PYTHON2_INCLUDE_DIR ${PYTHON_INCLUDE_DIR} CACHE PATH "include dir")
+    set(PYTHON2_DEBUG_LIBRARIES ${PYTHON_DEBUG_LIBRARIES} CACHE PATH "include debug libraries")
+    set(PYTHON2LIBS_VERSION_STRING ${PYTHONLIBS_VERSION_STRING} CACHE STRING "version")
```
Comment (mckellyln) [inline plugins/pyembed/CMakeLists.txt]: Any reason some of the string labels are different here than above ?
("Library" vs "library" for example)
("include debug libraries" vs "debug libraries for example")

File: plugins/pyembed/pyembed.cpp [hunk 6]
```diff
@@ -19,11 +19,20 @@
...
-#include "Python.h"
-#undef ssize_t
+  #define ssize_t python_ssize_t
+  #include "Python.h"
+  #undef ssize_t
 #else
-#include "Python.h"
+  #include "patchlevel.h"
+  #if (PY_MAJOR_VERSION == 2 && PY_MINOR_VERSION >= 7)
+    #ifdef _POSIX_C_SOURCE
```
Comment (mckellyln) [inline plugins/pyembed/pyembed.cpp]: Perhaps add a comment here such as "This Python version sets _POSIX_C_SOURCE and _XOPEN_SOURCE so we need to undef them here to prevent ...."

File: plugins/pyembed/CMakeLists.txt [hunk 7]
```diff
@@ -37,13 +37,14 @@ if(PY2EMBED)
...
     unset(PYTHONLIBS_VERSION_STRING CACHE)
     ADD_PLUGIN(py2embed PACKAGES PythonLibs MINVERSION 2.6 MAXVERSION 2.7)
-    set(PYTHON2LIBS_FOUND ${PYTHONLIBS_FOUND})
-    set(PYTHON2_LIBRARY ${PYTHON_LIBRARY})
-    set(PYTHON2_LIBRARIES ${PYTHON_LIBRARIES})
-    set(PYTHON2_INCLUDE_PATH ${PYTHON_INCLUDE_PATH})
-    set(PYTHON2_INCLUDE_DIR ${PYTHON_INCLUDE_DIR})
-    set(PYTHON2_DEBUG_LIBRARIES ${PYTHON_DEBUG_LIBRARIES})
-    set(PYTHON2LIBS_VERSION_STRING ${PYTHONLIBS_VERSION_STRING})
+
```
Comment (mckellyln) [inline plugins/pyembed/CMakeLists.txt]: remove added blank line

**Comment by mckellyln** [comment]
@Michael-Gardner a few comments from me and @richardkchapman

**Comment by mckellyln** [comment]
@Michael-Gardner what is the problem/error if _POSIX_C_SOURCE and _XOPEN_SOURCE are not undef'ed ?

**Comment by Michael-Gardner** [comment]
@mckellyln this is on a centos 6 box using 2.7

In file included from /opt/rh/python27/root/usr/include/python2.7/pyconfig.h:6:0,
                 from /opt/rh/python27/root/usr/include/python2.7/Python.h:8,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/pyembed.cpp:38:
/opt/rh/python27/root/usr/include/python2.7/pyconfig-64.h:1199:0: warning: "_POSIX_C_SOURCE" redefined [enabled by default]
 #define _POSIX_C_SOURCE 200112L
 ^
In file included from /usr/include/endian.h:22:0,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/./../../system/include/platform.h:79,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/pyembed.cpp:18:
/usr/include/features.h:162:0: note: this is the location of the previous definition
 # define _POSIX_C_SOURCE 200809L
 ^
In file included from /opt/rh/python27/root/usr/include/python2.7/pyconfig.h:6:0,
                 from /opt/rh/python27/root/usr/include/python2.7/Python.h:8,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/pyembed.cpp:38:
/opt/rh/python27/root/usr/include/python2.7/pyconfig-64.h:1221:0: warning: "_XOPEN_SOURCE" redefined [enabled by default]
 #define _XOPEN_SOURCE 600
 ^
In file included from /usr/include/endian.h:22:0,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/./../../system/include/platform.h:79,
                 from /home/centos/HPCC_Builds/HPCC-Platform/plugins/pyembed/pyembed.cpp:18:
/usr/include/features.h:164:0: note: this is the location of the previous definition
 # define _XOPEN_SOURCE 700

File: plugins/py3embed/CMakeLists.txt [hunk 8]
```diff
@@ -33,13 +33,13 @@ if(PY3EMBED)
...
     unset(PYTHONLIBS_VERSION_STRING CACHE)
     ADD_PLUGIN(py3embed PACKAGES PythonLibs MINVERSION 3.3)
-    set(PYTHON3LIBS_FOUND ${PYTHONLIBS_FOUND})
-    set(PYTHON3_LIBRARY ${PYTHON_LIBRARY})
-    set(PYTHON3_LIBRARIES ${PYTHON_LIBRARIES})
-    set(PYTHON3_INCLUDE_PATH ${PYTHON_INCLUDE_PATH})
-    set(PYTHON3_INCLUDE_DIR ${PYTHON_INCLUDE_DIR})
-    set(PYTHON3_DEBUG_LIBRARIES ${PYTHON_DEBUG_LIBRARIES})
-    set(PYTHON3LIBS_VERSION_STRING ${PYTHONLIBS_VERSION_STRING})
+    set(PYTHON3LIBS_FOUND ${PYTHONLIBS_FOUND} CACHE BOOL "Found python3 Libraries")
```
Comment (mckellyln) [inline plugins/py3embed/CMakeLists.txt]: Python3 to match capital P in Python2 below.

File: plugins/pyembed/pyembed.cpp [hunk 9]
```diff
@@ -19,11 +19,22 @@
...
-#define ssize_t python_ssize_t
-#include "Python.h"
-#undef ssize_t
+  #define ssize_t python_ssize_t
+  #include "Python.h"
+  #undef ssize_t
 #else
-#include "Python.h"
+  // This Python version Python.h sets _POSIX_C_SOURCE and _XOPEN_SOURCE so we
+  // need to undefine them to prevent redefinition warnings
```
Comment (mckellyln) [inline plugins/pyembed/pyembed.cpp]: Is it just a warning you are trying to eliminate ?  Or does the code need the newer values ?
If its just a warning then does it run ok without this ?
Would a pragma to disable warning be another way ?

Comment (Michael-Gardner) [reply plugins/pyembed/pyembed.cpp]: @xwang2713 has alluded to there being issues with some other third party libraries.  In my builds on centos 6 using python2.7 I've only run into warnings.  But that isn't to say that it isn't an issue on other platforms/library combinations.

Comment (mckellyln) [reply plugins/pyembed/pyembed.cpp]: ok, so OS defines _POSIX_C_SOURCE as 200809 and pyconfig-64.h redefines it as 200112
and OS defines _XOPEN_SOURCE as 700 abd pyconfig-64.h redefines it as 600
I wish pyconfig-64.h would honor these if they were already set to something >= what it wants.

Comment (mckellyln) [reply plugins/pyembed/pyembed.cpp]: @Michael-Gardner should the include "Python.h" happen before platform.h ?
Then platform.h undefs and defines so its ok.
Some Python documentation states that Python,h should be included before any system includes

**Comment by Michael-Gardner** [comment]
@mckellyln Made the final changes and squashed.

**Comment by mckellyln** [comment]
@Michael-Gardner ok, is there any testing that could be done ?

**Comment by Michael-Gardner** [comment]
Not that I know of.  John and his team are using it a fair bit internally so if there are issues we'll certainly hear about it.
